python/pipeline/utils/clocktools.py

```diff
@@ -0,0 +1,996 @@
+                f"{round(100*sum(source_nan_indices)/len(source_signal),2)}% of total signal. "
+                f"Largest NaN gap found: {round(biggest_time_gap, 2)} seconds."
+            )
+            print(msg)
+
+    ## Lowpass signal if needed
+    if target_fps < source_fps:
+        if debug:
+            msg = (
+                f"Source FPS of {round(source_fps,2)} is greater than target FPS {round(target_fps,2)}. "
+                f"Hamming lowpass filtering source signal before interpolation"
+            )
+            print(msg)
+        source_signal = shared.FilterMethod._lowpass_hamming(
+            signal=source_signal, signal_freq=source_fps, lowpass_freq=target_fps
+        )
+
+    ## Timing and recording array lengths can differ slightly if recording was stopped mid-scan. Timings for
+    ## the next X depths would be recorded, but fluorescence values would be dropped if all depths were not
+    ## recorded. This would mean timings difference shouldn't be more than the number of depths of the scan.
+    if len(source_times) < len(source_signal):
+        msg = (
+            f"More recording values than source time values exist! This should not be possible.\n"
+            f"Source time length: {len(source_times)}\n"
+            f"Source signal length: {len(source_signal)}"
+        )
+        raise PipelineException(msg)
+
+    elif len(source_times) > len(source_signal):
+
+        scan_res = pipe.ScanInfo.proj() & scan_key  ## To make sure we select all fields
+        z_plane_num = len(dj.U("z") & (pipe.ScanInfo.Field & scan_res))
+        if (len(source_times) - len(source_signal)) > z_plane_num:
+            msg = (
+                f"Extra timing values exceeds reasonable error bounds. "
+                f"Error length of {len(source_times) - len(source_signal)} with only {z_plane_num} z-planes."
+            )
+            raise PipelineException(msg)
+
+        else:
+            
+            shorter_length = np.min((len(source_times), len(source_signal)))
+            source_times = source_times[:shorter_length]
+            source_signal = source_signal[:shorter_length]
+            if debug:
+                length_diff = np.abs(len(source_times) - len(source_signal))
+                msg = (
+                    f"Source times and source signal show length mismatch within acceptable error."
+                    f"Difference of {length_diff} within acceptable bounds of {z_plane_num} z-planes."
+                )
+                print(msg)
+
+    ## Interpolating source signal into target timings
+    signal_interp = interpolate.interp1d(
+        source_times, source_signal, bounds_error=False
+    )
+    interpolated_signal = signal_interp(target_times)
+    if target_replace_nans is not None:
+        for target_nan_idx in target_replace_nans:
+            interpolated_signal[target_nan_idx] = np.nan
+
+    return interpolated_signal
+
```

Comment:
I think these four functions could be collapsed into a single one, where source format = ('index' , 'time') and desired target format = ('index' , 'time' , 'signal') was explicitly specified in the calling.

Reply:
I was also really confused that return_interpolate would result in returning source signal rather than target signal.  I would prefer that the output was always from the target class, otherwise the source target nomenclature becomes muddied very quickly.  If the source signal is required, the function can be called again with the roles reversed.  

Is there a nuanced use case where it makes sense to return the source signal in this way instead?